service/history/queues/action_pending_task_count.go

```diff
@@ -0,0 +1,226 @@
+	attributes *AlertAttributesQueuePendingTaskCount,
+	monitor Monitor,
+	maxReaderCount int,
+	completionFn actionCompletionFn,
+) Action {
+	return &actionQueuePendingTask{
+		attributes:     attributes,
+		monitor:        monitor,
+		maxReaderCount: maxReaderCount,
+		completionFn:   completionFn,
+	}
+}
+
+func (a *actionQueuePendingTask) Run(readerGroup *ReaderGroup) {
+	defer a.completionFn()
+
+	// first check if the alert is still valid
+	if a.monitor.GetTotalPendingTaskCount() <= a.attributes.CiriticalPendingTaskCount {
+		return
+	}
+
+	// then try to shrink existing slices, which may reduce pending task count
+	readers := readerGroup.Readers()
+	if a.tryShrinkSlice(readers) {
+		return
+	}
+
+	// have to unload pending tasks to reduce pending task count
+	a.init()
+	a.gatherStatistics(readers)
+	a.findSliceToClear(
+		a.monitor.GetTotalPendingTaskCount(),
+		int(float64(a.attributes.CiriticalPendingTaskCount)*targetThresholdCoefficient),
+	)
+	a.splitAndClearSlice(readers, readerGroup)
+}
+
+func (a *actionQueuePendingTask) tryShrinkSlice(
+	readers map[int32]Reader,
+) bool {
+	for _, reader := range readers {
+		reader.ShrinkSlices()
+	}
+	return a.monitor.GetTotalPendingTaskCount() <= a.attributes.CiriticalPendingTaskCount
+}
+
+func (a *actionQueuePendingTask) init() {
+	a.tasksPerNamespace = make(map[namespace.ID]int)
+	a.pendingTaskPerNamespacePerSlice = make(map[Slice]map[namespace.ID]int)
+	a.slicesPerNamespace = make(map[namespace.ID][]Slice)
+	a.namespaceToClearPerSlice = make(map[Slice][]namespace.ID)
+}
+
+func (a *actionQueuePendingTask) gatherStatistics(
+	readers map[int32]Reader,
+) {
+	// gather statistic for
+	// 1. total # of pending tasks per namespace
+	// 2. for each slice, # of pending taks per namespace
+	// 3. for each namespace, a list of slices that contains pending tasks from that namespace,
+	//    reversely ordered by slice range. Upon unloading, first unload newer slices.
+	for _, reader := range readers {
+		reader.WalkSlices(func(s Slice) {
+			a.pendingTaskPerNamespacePerSlice[s] = s.TaskStats().PendingPerNamespace
+			for namespaceID, pendingTaskCount := range a.pendingTaskPerNamespacePerSlice[s] {
+				a.tasksPerNamespace[namespaceID] += pendingTaskCount
+				a.slicesPerNamespace[namespaceID] = append(a.slicesPerNamespace[namespaceID], s)
+			}
+		})
+	}
+	for _, sliceList := range a.slicesPerNamespace {
+		slices.SortFunc(sliceList, func(this, that Slice) bool {
+			thisMin := this.Scope().Range.InclusiveMin
+			thatMin := that.Scope().Range.InclusiveMin
+			return thisMin.CompareTo(thatMin) > 0
+		})
+	}
+}
+
+func (a *actionQueuePendingTask) findSliceToClear(
+	currentPendingTasks int,
+	targetPendingTasks int,
+) {
+	// order namespace by # of pending tasks
+	pq := collection.NewPriorityQueue(func(this, that namespace.ID) bool {
+		return a.tasksPerNamespace[this] > a.tasksPerNamespace[that]
+	})
+	for namespaceID := range a.tasksPerNamespace {
+		pq.Add(namespaceID)
+	}
```

Comment:
random thought: this is NlogN time. Build heap could be built in O(N) time.

Reply:
👍 Added `NewPriorityQueueWithItems` to build heap in O(n) time. We still need the for loop for convert the map to a slice of namespace.ID though.